### PR TITLE
Ideal

### PR DESCRIPTION
**Pull request by msurmava**

### CODE REVIEW

File: optimal/Ideal.rb [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+    end
+
+    def validate_parameters
+      if @height == 0
+        warn "#{@name.capitalize}, you entered invalid weight"
+        exit 1
+      end
+    end
+
+    def optimal
```
Comment (cyberik-dev) [inline optimal/Ideal.rb:21]: let's call it 'call' 

Service objects are single business actions. You can change the name of your public method if you like. I prefer using call

File: optimal/Ideal.rb [hunk 2]
```diff
@@ -0,0 +1,33 @@
+#calculates ideal weight by height
+
+class Ideal_weight
```
[the remark this answers is not in the document]
Comment (msurmava) [reply optimal/Ideal.rb]: done

File: optimal/Ideal.rb [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+    EDGE  = 0
+
+    def initialize
+      puts  "Enter your name:"
+      @name = gets.chomp
+      puts  "Enter your height"
+      @height = gets.chomp.to_i
+    end
+
+    def validate_parameters
```
[the remark this answers is not in the document]
Comment (msurmava) [reply optimal/Ideal.rb]: done

File: optimal/Ideal.rb [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+      if @height == 0
```
[the remark this answers is not in the document]
Comment (msurmava) [reply optimal/Ideal.rb]: It returns message that user entered invalid height

File: optimal/Ideal.rb (hunk 1 above)
Comment (msurmava) [reply optimal/Ideal.rb]: done

File: optimal/Ideal.rb [hunk 5]
```diff
@@ -0,0 +1,33 @@
+#calculates ideal weight by height
+
+class Ideal_weight
+    TOSUBTRACT = 100
```
[the remark this answers is not in the document]
Comment (msurmava) [reply optimal/Ideal.rb:4]: I have constants inside of class

File: optimal/Ideal.rb [hunk 6]
```diff
@@ -0,0 +1,34 @@
+#calculates ideal weight by height
```
[the remark this answers is not in the document]
Comment (cyberik-dev) [reply optimal/Ideal.rb:1]: please do not capitalize the file name - Ideal.rb and this name should be corresponding with the class name

File: optimal/Ideal.rb [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+      if ideal  < EDGE
+        "#{@name.capitalize}, your weight is already optimal!"
+      else
+        "#{@name.capitalize}, your ideal weight is #{ideal} kgs!"
+      end
+    end
+
+    private
+    def validate_parameters
+      if @height == 0
```
[the remark this answers is not in the document]
Comment (cyberik-dev) [reply optimal/Ideal.rb:26]: change to `return unless @height <= 0`

File: optimal/Ideal.rb [hunk 8]
```diff
@@ -0,0 +1,34 @@
+#calculates ideal weight by height
+
+class IdealWeightCounter
+    TOSUBTRACT = 100
```
Comment (cyberik-dev) [inline optimal/Ideal.rb:4]: rename it to SUBTRACTOR

Comment (cyberik-dev) [reply optimal/Ideal.rb:4]: the value should be 110

File: optimal/Ideal.rb [hunk 9]
```diff
@@ -0,0 +1,34 @@
...
+    def initialize
+      puts  "Enter your name:"
+      @name = gets.chomp
+      puts  "Enter your height"
+      @height = gets.chomp.to_i
+    end
+
+    def call
+      validate_parameters
+      ideal = @height - TOSUBTRACT
```
[the remark this answers is not in the document]
Comment (cyberik-dev) [reply optimal/Ideal.rb:16]: and move to a method
```ruby
  def ideal_weight
    @ideal_weight ||= @height - SUBTRACTOR
  end
```

File: optimal/Ideal.rb [hunk 10]
```diff
@@ -0,0 +1,34 @@
...
+      if ideal  < EDGE
```
Comment (cyberik-dev) [inline optimal/Ideal.rb:17]: move to 

```ruby
  def ideal_weight?
    ideal_weight < EDGE
  end
```

**Review by cyberik-dev — CHANGES_REQUESTED** [review]

File: optimal/Ideal.rb [hunk 11]
```diff
@@ -0,0 +1,34 @@
...
+        "#{@name.capitalize}, your weight is already optimal!"
+      else
+        "#{@name.capitalize}, your ideal weight is #{ideal} kgs!"
+      end
+    end
+
+    private
```
[the remark this answers is not in the document]
Comment (msurmava) [reply optimal/Ideal.rb]: done

File: optimal/Ideal.rb (hunk 7 above)
[the remark this answers is not in the document]
Comment (msurmava) [reply optimal/Ideal.rb]: done

File: optimal/Ideal.rb (hunk 8 above)
Comment (msurmava) [reply optimal/Ideal.rb]: corrected

File: optimal/Ideal.rb (hunk 10 above)
Comment (msurmava) [reply optimal/Ideal.rb]: done

File: idealweightcounter.rb [hunk 12]
```diff
@@ -0,0 +1,48 @@
...
+    validate_parameters
+    calculate_ideal_weight
+    ideal_weight? ? already_opimal_weight_message : ideal_weight_message
+  rescue AttributeError => e
+    warn e
+  end
+
+  private
+
+  def calculate_ideal_weight
```
Comment (cyberik-dev) [inline idealweightcounter.rb:25]: please read about Ruby Memoization 
https://www.honeybadger.io/blog/ruby-rails-memoization/
https://www.justinweiss.com/articles/4-simple-memoization-patterns-in-ruby-and-one-gem/
https://thoughtbot.com/blog/ruby-memoization-and-alternatives

and use it here

Comment (cyberik-dev) [inline idealweightcounter.rb:25]: rename to ideal_weight and call this method instead of using variable @ideal

**Review by cyberik-dev — CHANGES_REQUESTED** [review]

File: idealweightcounter.rb [hunk 13]
```diff
@@ -0,0 +1,48 @@
+# calculates ideal weight by height
```
Comment (cyberik-dev) [inline idealweightcounter.rb:1]: rubocop 

1 file inspected, 2 offenses detected, 2 offenses auto-correctable

Comment (msurmava) [reply idealweightcounter.rb:3]: no more offenses

**Pull request merged**